File: 2023/src/bin/template/main.rs
use std::borrow::Cow;

use advent_2023::{BasicSolution, Solution};
use winnow::{combinator::rest, Parser};

struct Day;

impl BasicSolution for Day {
    const INPUT: &'static str = include_str!("data.txt");
    const SAMPLE_INPUT: &'static str = include_str!("sample.txt");

    type Shared = Vec<&'static str>;
    type Answer = u32;

    const SAMPLE_ANSWER_A: Self::TestAnswer = 0;
    const SAMPLE_ANSWER_B: Self::TestAnswer = 0;

    fn shared(input: &'static str) -> anyhow::Result<Self::Shared> {
        input
            .lines()
            .map(|line| line_parser.parse(line).map_err(anyhow::Error::msg))
            .collect()
    }

    fn part_a(shared: Cow<Self::Shared>) -> anyhow::Result<Self::Answer> {
        todo!("{shared:?}")
    }

    fn part_b(_: Self::Shared) -> anyhow::Result<Self::Answer> {
        Ok(0)
    }
}

fn line_parser(input: &mut &'static str) -> winnow::PResult<&'static str> {
    rest.parse_next(input)
}

fn main() -> anyhow::Result<()> {
    Day::main()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a() -> anyhow::Result<()> {
        Day::test_part_a()
    }

    #[test]
    fn b() -> anyhow::Result<()> {
        Day::test_part_b()
    }
}

File: 2023/python/template/main.py
import collections  # noqa: F401
import itertools  # noqa: F401
import math  # noqa: F401
import re  # noqa: F401
from dataclasses import dataclass  # noqa: F401
from pathlib import Path

SAMPLE_INPUT = """
"""
SAMPLE_INPUT_B = SAMPLE_INPUT
INPUT = Path("data.txt").read_text()


def part_a(input: str):
    return 0


def part_b(input: str):
    return 0


def test_part_a(capsys):
    assert part_a(SAMPLE_INPUT) == 0
    with capsys.disabled():
        print(f"Part A: {part_a(INPUT)}")


def test_part_b(capsys):
    assert part_b(SAMPLE_INPUT_B) == 0
    with capsys.disabled():
        print(f"Part B: {part_b(INPUT)}")
